File: sem iii/NM/Nm lab/bisection_lab.c
#include <stdio.h>
#include <math.h>

// f(x) = x^2 - 4 sin(x)
double fx(double x) {
    return x*x - 4*sin(x);
}

int main() {
    double xl, xu, xm, xm_old, eps;
    int itr, i;

    printf("Program to find root using Bisection Method\n");
    printf("Enter the value of x_l (lower bracket): ");
    scanf("%lf", &xl);
    printf("Enter the value of x_u (upper bracket): ");
    scanf("%lf", &xu);
    printf("Enter the error tolerance: ");
    scanf("%lf", &eps);
    printf("Enter the number of iterations: ");
    scanf("%d", &itr);

    // Check if root exists
    if (fx(xl) * fx(xu) > 0) {
        printf("Root does not exist in the given interval.\n");
        return 0;
    }

    xm_old = xl;

    for (i = 0; i < itr; i++) {
        xm = (xl + xu) / 2.0;

        if (fx(xl) * fx(xm) < 0)
            xu = xm;
        else
            xl = xm;

        double error = fabs(xm - xm_old);

        printf("Iteration %d: xl=%.3lf xu=%.3lf xm=%.3lf error=%.3lf\n",
               i+1, xl, xu, xm, error);

        if (error < eps)
            break;

        xm_old = xm;
    }

    printf("\nThe value of the root is %.3lf\n", xm);
    return 0;
}

File: sem iii/NM/Nm lab/ho/bisection_lab.py
import math

# Lab 1 - Bisection Method
# f(x) = x^2 - 4 sin(x)

def fx(x):
    return x**2 - 4*math.sin(x)

def bisection(xu, xl, eps, itr):
    xm_old = xl  # for error calculation

    for i in range(itr):
        f_u = fx(xu)
        f_l = fx(xl)

        # Check if root exists
        if f_u * f_l > 0:
            return "Root does not exist in the initial guess values. Please try other guesses."

        # Midpoint
        xm = (xu + xl) / 2
        f_m = fx(xm)

        # Decide the new interval
        if f_l * f_m < 0:
            xu = xm
        else:
            xl = xm

        # Error calculation
        error = abs(xm - xm_old)

        print(f"The error in the {i+1} iteration is: {error:.3f} "
              f"and xu={xu:.3f}, xl={xl:.3f} and xm={xm:.3f}.")

        # Check stopping condition
        if error < eps:
            return xm

        xm_old = xm

    return xm


if __name__ == "__main__":
    print("Program to return the root of the function using the bisection method.")
    xl = float(input("Enter the value of x_l (lower bracket): "))
    xu = float(input("Enter the value of x_u (upper bracket): "))
    eps = float(input("Enter the value of error tolerance: "))
    itr = int(input("Enter the number of iterations: "))

    root = bisection(xu, xl, eps, itr)

    if isinstance(root, str):
        print(root)
    else:
        print(f"\nThe value of the root is {root:.3f}")
